### PR TITLE
DOC: Improved documentation for jax.numpy.roll

### PR DESCRIPTION
**Pull request by selamw1**
Part of https://github.com/google/jax/issues/21461

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -7807,9 +7807,53 @@ def _roll_static(a: Array, shift: Sequence[int], axis: Sequence[int]) -> Array:
...
+      all axes are shifted by the same amount. If a tuple, the shift for each
+      axis is specified individually.
+    axis: the axis or axes to roll. If ``None``, the array is flattened, shifted,
+      and then reshaped to its original shape.
+
+  Returns:
+    A copy of ``a`` with elements rolled along the specified axis or axes.
+
+  Notes:
+    Unlike :func:`numpy.roll`, :func:`jax.numpy.roll` will return a copy rather
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7828]: I don't think this note is correct to include here: `np.roll` returns a copy, not a view.

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -7807,9 +7807,53 @@ def _roll_static(a: Array, shift: Sequence[int], axis: Sequence[int]) -> Array:
...
+    than a view of the input array. However, under JIT, the compiler will optimize away
+    such copies when possible, so this doesn't have performance impacts in practice.
+
+  See also:
+    - :func:`jax.numpy.rollaxis`: roll the specified axis to a given position.
+    - :func:`jax.numpy.moveaxis`: move an array axis to a new position.
+    - :func:`jax.numpy.swapaxes`: swap two axes.
+    - :func:`jax.numpy.transpose`: general permutation of axes.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7836]: I'd only include `rollaxis` here. The rest are transpose-like operations and not particularly related to `roll`

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -7807,9 +7807,53 @@ def _roll_static(a: Array, shift: Sequence[int], axis: Sequence[int]) -> Array:
...
+    Roll elements along a specific axis:
+
+    >>> a = jnp.array([[ 0,  1,  2,  3],
+    ...                [ 4,  5,  6,  7],
+    ...                [ 8,  9, 10, 11]], dtype=int32)
+    >>> jnp.roll(a, 1, axis=0)
+    Array([[ 8,  9, 10, 11],
+           [ 0,  1,  2,  3],
+           [ 4,  5,  6,  7]], dtype=int32)
+    >>> jnp.roll(a, [3, 3], axis=1)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7852]: Can we use (2, 3) here instead? Rolling a size-3 axis 3 steps results in no change.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks! Some comments below

**Comment by selamw1** [comment]
Thanks @jakevdp for a quick review!
Comments fixed.

File: jax/_src/numpy/lax_numpy.py [hunk 4]
```diff
@@ -7807,9 +7807,45 @@ def _roll_static(a: Array, shift: Sequence[int], axis: Sequence[int]) -> Array:
...
+    ...                [ 4,  5,  6,  7],
+    ...                [ 8,  9, 10, 11]], dtype=int32)
+    >>> jnp.roll(a, 1, axis=0)
+    Array([[ 8,  9, 10, 11],
+           [ 0,  1,  2,  3],
+           [ 4,  5,  6,  7]], dtype=int32)
+    >>> jnp.roll(a, [2, 3], axis=1)
+    Array([[ 3,  0,  1,  2],
+           [ 7,  4,  5,  6],
+           [11,  8,  9, 10]], dtype=int32)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7847]: This output looks incorrect to me - I guess the doctests will tell 😁

File: jax/_src/numpy/lax_numpy.py [hunk 5]
```diff
@@ -7807,9 +7807,45 @@ def _roll_static(a: Array, shift: Sequence[int], axis: Sequence[int]) -> Array:
...
+    Roll elements along a specific axis:
+
+    >>> a = jnp.array([[ 0,  1,  2,  3],
+    ...                [ 4,  5,  6,  7],
+    ...                [ 8,  9, 10, 11]], dtype=int32)
+    >>> jnp.roll(a, 1, axis=0)
+    Array([[ 8,  9, 10, 11],
+           [ 0,  1,  2,  3],
+           [ 4,  5,  6,  7]], dtype=int32)
+    >>> jnp.roll(a, [2, 3], axis=1)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7844]: Remove `axis=1` here? Then it will roll separately along the two dimensions

Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py:7844]: how about `jnp.roll(a, [2, 3], axis=0)` ? since it's one of the examples for 'Roll elements along a specific axis'.

File: jax/_src/numpy/lax_numpy.py (hunk 4 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py:7847]: It seems the doctests passed :smile: , however, i should remove `, dtype=int32` from below array:
```
    >>> a = jnp.array([[ 0,  1,  2,  3],
    ...                [ 4,  5,  6,  7],
    ...                [ 8,  9, 10, 11]], dtype=int32)
```

File: jax/_src/numpy/lax_numpy.py (hunk 5 above)
Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:7844]: What's the intended behavior here? Roll 2 followed by roll 3? So `jnp.roll(a, [2, 3], axis=0)` is equivalent to `jnp.roll(a, 5, axis=0)`? That's really counter-intuitive, and is probably more an implementation detail than and intented use of the API.

I think a more useful example would be to show that you can roll `2` along the first axis, and `3` along the second axis: `jnp.roll(a, (2, 3), axis=(0, 1))`

File: jax/_src/numpy/lax_numpy.py (hunk 4 above)
Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:7847]: Yeah, `int32` isn't a defined variable in this context.

**Review by jakevdp — APPROVED** [review]
Looks good, thanks!
